students/smandava/session6/sm_mailroom_refactor.py

```
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+"""Nov 27 2016 Mail room Exercise refactoring for testing."""
+import operator
+donor_dict = {'Alfred Hitchcock': [10, 20, 30],
+              'Martin Scorsese': [20, 30, 40],
+              'Steven Spielberg': [40, 20, 30, 10],
+              'Francis Coppola': [10, 10, 10],
+              'Ridley Scott': [10, 20, 30, 10]}
+
+donor_summary_dict = {}
+donor_report = []
+
+def print_report():
+    report_header = ("{0:<30s} {1:<30s} {2:<30s} {3:<30s}".
+                     format('Donor Name',
+                            'Total Amount', 'No.of Donations',
+                            'Avg.Donation'))
+
+
+    donor_summary_dict = {donorname: [sum(donor_dict[donorname]), len(donor_dict[donorname]), (sum(donor_dict[donorname]) / len(donor_dict[donorname]))] for donorname in donor_dict.keys()}
+    sorted_donors = sorted(donor_summary_dict.items(),
+                           key=operator.itemgetter(1), reverse=True)
+    for donor in (sorted_donors):
+        donor_report.append(donor[0] + ' ' + str(donor[1][0]) + ' ' + str(donor[1][1]) + ' ' + str(donor[1][2]))
+    return "\n".join(donor_report)
+
+
+def send_thanks(user_input, d_amount=None):
+    if (user_input == 'list'):
+        return(donor_dict)
+    if user_input in donor_dict.keys():
+        try:
+            d_amount = int(d_amount)
+        except ValueError as error:
+            return (type(error).__name__)
+        donor_dict.setdefault(user_input, []).append(d_amount)
+        return (user_input, donor_dict.get(user_input))
+    else:
+        try:
+            d_amount = int(d_amount)
+        except ValueError as error:
+            return (type(error).__name__)
```

why return the error like this?

what you want is to re-prompt the user for a valid number. If you aren't going to do that in this function, then you can let the exception be raised, and catch it where it is called.